File: fixtures/ts-sample/src/closures.ts
// cgx-fixture: closures and higher-order functions (TypeScript)
// Covers: calls:closure (arrow function invocation), calls:callback (fn-value arg),
//         closure captures, .map/.filter/.reduce iterator callbacks

/// Arrow function stored in variable and invoked — calls:closure.
export function closureVariable(): number {
    const double = (x: number) => x * 2;
    return double(5);
}

/// .map() with arrow function — calls:closure on map body.
export function doubleAll(items: number[]): number[] {
    return items.map(x => x * 2);
}

/// .filter() — calls:closure (predicate).
export function filterEvens(items: number[]): number[] {
    return items.filter(x => x % 2 === 0);
}

/// .reduce() — calls:closure (reducer).
export function sumArray(items: number[]): number {
    return items.reduce((acc, x) => acc + x, 0);
}

/// Closure that captures outer variable — captured binding.
export function makeAdder(n: number): (x: number) => number {
    return (x: number) => x + n;  // captures n
}

export function useCapturedClosure(): number {
    const add5 = makeAdder(5);
    return add5(10);
}

/// Higher-order function — calls:callback on f(value).
export function apply(f: (x: number) => number, value: number): number {
    return f(value);
}

export function useCallback(): number {
    return apply(x => x + 1, 41);  // calls:callback (arrow fn as argument)
}

/// Function reference as argument — calls:callback.
function negate(x: number): number { return -x; }

export function transformList(items: number[], f: (x: number) => number): number[] {
    return items.map(f);  // calls:callback on f inside map
}

export function negateList(items: number[]): number[] {
    return transformList(items, negate);  // negate passed as callback
}

/// Nested closures.
export function nestedClosures(): number {
    const outer = (x: number) => {
        const inner = (y: number) => y * y;
        return inner(x) + x;
    };
    return outer(3);
}

/// Method reference as callback — calls:callback.
export class Formatter {
    prefix: string;

    constructor(prefix: string) {
        this.prefix = prefix;
    }

    format(s: string): string {
        return `${this.prefix}${s}`;
    }
}

export function formatAll(items: string[], fmt: Formatter): string[] {
    return items.map(s => fmt.format(s));  // calls:closure with captured fmt
}

File: fixtures/rust-sample/src/closures.rs
// cgx-fixture: closures and iterators
// Covers: calls:closure (closure invocation), calls:callback (fn-value arg),
//         implicit:iterator (for loop over iterator), closure captures

/// Closure assigned to a variable and invoked — calls:closure.
pub fn closure_variable() -> i32 {
    let double = |x: i32| x * 2;
    double(5)
}

/// Iterator body — implicit:iterator on the for loop's `next()` calls.
pub fn sum_vec(items: &[i32]) -> i32 {
    let mut total = 0;
    for item in items.iter() {
        total += item;
    }
    total
}

/// .map() with a closure — calls:closure on the map body.
pub fn double_all(items: &[i32]) -> Vec<i32> {
    items.iter().map(|x| x * 2).collect()
}

/// Closure that captures a variable from outer scope.
pub fn make_adder(n: i32) -> impl Fn(i32) -> i32 {
    move |x| x + n
}

pub fn use_captured_closure() -> i32 {
    let add5 = make_adder(5);
    add5(10)
}

/// Higher-order function — calls:callback on invocation of `f`.
pub fn apply<F: Fn(i32) -> i32>(f: F, value: i32) -> i32 {
    f(value)
}

pub fn use_callback() -> i32 {
    apply(|x| x + 1, 41)
}

/// Closure passed as a function pointer — calls:callback.
pub fn transform_list(items: &mut Vec<i32>, f: fn(i32) -> i32) {
    for item in items.iter_mut() {
        *item = f(*item);
    }
}

fn negate(x: i32) -> i32 {
    -x
}

pub fn negate_list(items: &mut Vec<i32>) {
    transform_list(items, negate);
}

/// Nested closures — closure capturing another closure.
pub fn nested_closures() -> i32 {
    let outer = |x: i32| {
        let inner = |y: i32| y * y;
        inner(x) + x
    };
    outer(3)
}

/// .filter() + .map() chain — multiple closure invocations.
pub fn filter_and_double(items: &[i32], threshold: i32) -> Vec<i32> {
    items.iter()
        .filter(|&&x| x > threshold)
        .map(|&x| x * 2)
        .collect()
}
